File: 1/2021.03.27/652.Find Duplicate Subtrees/java/solution.java
class Solution {
    Map<String, Integer> map = new HashMap<>();
    
    List<TreeNode> res = new ArrayList<>();
    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        post(root);
        return res;
    }
    
    private String post(TreeNode root) {
        if (root == null) {
            return "#";
        }
        String s = post(root.left) + "," + post(root.right) + "," + root.val;
        if (map.getOrDefault(s, 0) == 1) {
            res.add(root);
        }
        map.put(s, map.getOrDefault(s, 0) + 1);
        return s;
    }
}

File: 1/2021.03.27/652.Find Duplicate Subtrees/python/solution.py
class Solution:
    def findDuplicateSubtrees(self, root: TreeNode) -> List[TreeNode]:
        def getid(root):
            if root:
                id = treeid[root.val, getid(root.left), getid(root.right)]
                trees[id].append(root)
                return id
        trees = collections.defaultdict(list)
        treeid = collections.defaultdict()
        treeid.default_factory = treeid.__len__
        getid(root)
        return [roots[0] for roots in trees.values() if roots[1:]]
